File: scripts/tfalv2_get_lux.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
File:
tfrgbledv2_set_color.py
Project:
Domoticz-Home-Automation-Workbook
Purpose:
Getter Example - Test getting the lux value of a Tinkerforge Ambient Light Bricklet 2.0.
Depends on: Python API Binding
Description:
JSON formatted string as argument with following key:value pairs:
IMPORTANT: all keys must be in uppercase. all values are string (in "").
Key:Value
"HOST":"String" - IP address of the host running the brick deamon, i.e. localhost or any other IP address
"PORT":integer - Port number. Default: 4223
"UID":"String" - UID of the bricklet. Get from the Brick Viewer, i.e. "yyc"
Arguments: '{"HOST":"localost","PORT":4223,"UID":"yyc"}'
At least 1 argument must be given.
Run example to get the lux value of the bricklet connected to the master brick which is connected via USB with Raspberry Pi:
cd domoticz/scripts/python
python3 /home/pi/domoticz/scripts/python/tfalv2_get_lux.py '{"UID":"yyc"}'
Output:
{"status": "OK", "title": "", "lux": "177.92"}
Version:
20200224 by rwbL
"""
import sys
import json

# Tinkerforge
from tinkerforge.ip_connection import IPConnection
from tinkerforge.bricklet_ambient_light_v2 import BrickletAmbientLightV2
## Set defaults for connection and RGB 
HOST = "localhost"
PORT = 4223
UID = "yyc"

# Arguments Length
ARGVLEN = 2

if __name__ == "__main__":
    status = 1
    # Get the command line argurments
    # print("Arguments: {}".format(str(sys.argv)))
    if len(sys.argv) == ARGVLEN:
        # Parse the command line argument 1 from string to json 
        args = json.loads(sys.argv[1])
        if "HOST" in args:
            HOST = args["HOST"]
        if "PORT" in args:
            PORT = int(args["PORT"])
        if "UID" in args:
            UID = args["UID"]
        # Check arguments
        # print("JSON:H={},P={},U={}".format(HOST,PORT,UID))
    else:
        status = 0
        result = { "status" : "ERROR", "title" : "Wrong number of arguments." }

    if status == 1:
        try:
            # Create IP connection
            ipcon = IPConnection()
            # Create device object
            al = BrickletAmbientLightV2(UID, ipcon)
            # Connect to brickd
            ipcon.connect(HOST, PORT)
            # Check if the bricklet is reachable - else error is triggered
            # In case the bricklet is not reacable, it takes 1-2 seconds befor a response is given
            al.get_identity()
            # # Get current illuminance
            illuminance = al.get_illuminance()
            if illuminance > 0:
                illuminance = illuminance/100.0
            result = { "status" : "OK", "title" : "", "lux" : "{}".format(illuminance) }
            # Disconnect
            ipcon.disconnect()
        except:
            result = { "status" : "ERROR", "title" : "get_illuminance failed. Check master & bricklet." }
    # Print the result which is used by the dzVents Lua script
    print(json.dumps(result))
 

File: scripts/tfalv2_get_lux.lua
-- Domoticz Home Automation Workbook - Function Tinkerforge
-- dzVents Automation Script: tfalv2_get_lux_python
-- Getter Example
-- The script is triggered every minute and executes python script.
-- The python script expects an argument with key:value pairs:
-- '{"HOST":"IP-Address","PORT":4223,"UID":"yyc"}'
-- At least 1 argument is mandatory.
-- The result of the command is a JSON formatted string, i.e.
-- {"status": "OK", "title": "", "lux":"146.21"}
-- {"status": "ERROR", "title": "Error message"}
-- VirtualDevice
-- Type:Lux, SubType:Lux, Name: MakeLab Lux, Idx: 106
-- 20200224 by rwbL

-- json library
local JSON = (loadfile "/home/pi/domoticz/scripts/lua/JSON.lua")()  -- For Linux

local CMD = "python3 /home/pi/domoticz/scripts/python/tfalv2_get_lux.py"

-- Idx of the devices
local IDXLUX = 107
-- TF Bricklet parameter
local TFUID = "yyc"

-- run external command
local function osExecute(cmd)
    local fileHandle     = assert(io.popen(cmd, 'r'))
    local commandOutput  = assert(fileHandle:read('*a'))
    local returnTable    = {fileHandle:close()}
    -- rc[3] contains returnCode
    return commandOutput,returnTable[3]
end

return {
	on = {
		timer = {
			'every minute',
	   },
    },
	execute = function(domoticz, timer)
        -- define the python command with json argument
        -- example: python3 tfalv2_get_lux.py '{"UID":"yyc"}'
        local pycmd = string.format('%s \'{\"UID\":"%s"}\'',CMD,TFUID)
		domoticz.log(pycmd, domoticz.LOG_INFO)
		
		-- run the external python scripts ad handle return code
		local output,rt = osExecute(pycmd)
        -- domoticz.log(string.format('%d:%s',rt,output), domoticz.LOG_INFO)
        -- decode the result string which is json format, i.e. 
        -- {"status": "OK", "title": "", "lux":"146.21"}
        -- {"status": "ERROR", "title": "Error message"}
        local result = JSON:decode(output);
        -- domoticz.log(string.format('TEST Status:%s, Title:%s, Lux:%s',result.status,result.title,result.lux), domoticz.LOG_INFO)
		-- handle the result of the python command: 0=OK,>0=ERROR
		if rt == 0 then
		    -- handle the result of the tinkerforge scripts
		    if result.status == "OK" then
                domoticz.log(string.format('Status:%s, Title:%s, Lux:%s',result.status,result.title,result.lux), domoticz.LOG_INFO)
                domoticz.devices(IDXLUX).updateLux(result.lux)
            -- tinkerforge script returns an error
            else
                domoticz.log(string.format('Status:%s, Title:%s, Lux:%s',result.status,result.title,result.lux), domoticz.LOG_ERROR)
            end
		-- handle the result of the python command
        else
            domoticz.log(string.format('Python command: RT:%d, Output:%s',rt,output), domoticz.LOG_ERROR)
        end

    end
}
